python/openai/openai_frontend/engine/triton_engine.py

```diff
@@ -0,0 +1,419 @@
+# EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+# PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+# PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+# OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+
+from __future__ import annotations
+
+import os
+import time
+import uuid
+from dataclasses import dataclass
+from typing import Any, Callable, Dict, Iterator, List, Optional
+
+import tritonserver
+from engine.engine import OpenAIEngine
+from schemas.openai import (
+    ChatCompletionChoice,
+    ChatCompletionFinishReason,
+    ChatCompletionResponseMessage,
+    ChatCompletionStreamingResponseChoice,
+    ChatCompletionStreamResponseDelta,
+    Choice,
+    CreateChatCompletionRequest,
+    CreateChatCompletionResponse,
+    CreateChatCompletionStreamResponse,
+    CreateCompletionRequest,
+    CreateCompletionResponse,
+    FinishReason,
+    Model,
+    ObjectType,
+)
+from utils.tokenizer import get_tokenizer
+from utils.triton import (
+    _create_trtllm_inference_request,
+    _create_vllm_inference_request,
+    _get_output,
+    _validate_triton_responses_non_streaming,
+)
+
+
+# TODO: Improve type hints
+@dataclass
+class TritonModelMetadata:
+    # Name used in Triton model repository
+    name: str
+    # Name of backend used by Triton
+    backend: str
+    # Triton model object handle
+    model: tritonserver.Model
+    # Tokenizers used for chat templates
+    tokenizer: Optional[Any]
+    # Time that model was loaded by Triton
+    create_time: int
+    # Conversion format between OpenAI and Triton requests
+    request_converter: Callable
+
+
+class TritonOpenAIEngine(OpenAIEngine):
+    def __init__(self, server: tritonserver.Server):
+        # Assume an already configured and started server
+        self.server = server
+
+        # NOTE: Creation time and model metadata will be static at startup for
+        # now, and won't account for dynamically loading/unloading models.
+        self.create_time = int(time.time())
+        self.model_metadata = self._get_model_metadata()
+
+    def ready(self) -> bool:
+        return self.server.ready()
+
+    def metrics(self) -> str:
+        return self.server.metrics()
+
+    def models(self) -> List[Model]:
+        models = []
+        for metadata in self.model_metadata.values():
+            models.append(
+                Model(
+                    id=metadata.name,
+                    created=metadata.create_time,
+                    object=ObjectType.model,
+                    owned_by="Triton Inference Server",
+                ),
+            )
+
+        return models
+
+    def chat(
+        self, request: CreateChatCompletionRequest
+    ) -> CreateChatCompletionResponse | Iterator[str]:
```

Comment:
Note: this should probably return `Iterator[CreateChatCompletionResponse]` instead of `Iterator[str]` for the streaming case, but would need to somehow wrap the generator to do the conversion from Object to SSE string on frontend side for each yielded item.

Same for completions.